panic: emitter is closed

<!--
You are about to open an issue in the ipfs-cluster repository. Please verify that:

* This is not a IPFS Cluster website content issue: file those here: https://github.com/ipfs/ipfs-cluster-website/issues
* You read the troubleshooting docs (https://cluster.ipfs.io/documentation/troubleshooting/)
* You searched for similar issues in the repo without luck
* All peers are running the same cluster version
* All peers are configured using the same cluster `secret`

Thank you!
-->

**Additional information:**
 - OS: Linux
 - IPFS Cluster version:  master, ef8f2cb17d03f309deb534227e565a3025a9d3cf
 - Installation method: built from source

**Describe the bug:**
 
My port for rest api was occupied, when I ran tests `go test -v ./api/rest`. Test failed as expected. I freed the port. On running the test again I saw this error

```
➜  ipfs-cluster ✗ go test -v ./api/rest 
=== RUN   TestLoadEmptyJSON
--- PASS: TestLoadEmptyJSON (0.00s)
=== RUN   TestLoadJSON
--- PASS: TestLoadJSON (0.00s)
=== RUN   TestApplyEnvVars
--- PASS: TestApplyEnvVars (0.00s)
=== RUN   TestLibp2pConfig
panic: emitter is closed

goroutine 71 [running]:
github.com/libp2p/go-eventbus.(*emitter).Emit(0xc000191200, 0xc81540, 0xc0000b7980)
	/home/kishan/code/pkg/mod/github.com/libp2p/go-eventbus@v0.0.3/basic.go:33 +0x6f
github.com/libp2p/go-libp2p/p2p/host/basic.(*BasicHost).RemoveStreamHandler(0xc0000d62c0, 0xd4501d, 0xc)
	/home/kishan/code/pkg/mod/github.com/libp2p/go-libp2p@v0.2.1/p2p/host/basic/basic_host.go:419 +0x112
github.com/libp2p/go-libp2p-gostream.(*listener).Close(0xc000046400, 0xc00014add0, 0x10)
	/home/kishan/code/pkg/mod/github.com/libp2p/go-libp2p-gostream@v0.1.2/listener.go:39 +0x53
net/http.(*onceCloseListener).close(...)
	/usr/local/go/src/net/http/server.go:3286
sync.(*Once).Do(0xc0000b7570, 0xc00014ae00)
	/usr/local/go/src/sync/once.go:44 +0xb3
net/http.(*onceCloseListener).Close(0xc0000b7560, 0x0, 0x55)
	/usr/local/go/src/net/http/server.go:3282 +0x54
net/http.(*Server).Serve(0xc0003684e0, 0xf91f80, 0xc000046400, 0xf815e0, 0xc0000b2210)
	/usr/local/go/src/net/http/server.go:2879 +0x4b0
github.com/ipfs/ipfs-cluster/api/rest.(*API).runLibp2pServer(0xc00012b050, 0xf95700, 0xc000362980)
	/home/kishan/code/ipfs-cluster/api/rest/restapi.go:477 +0x322
created by github.com/ipfs/ipfs-cluster/api/rest.(*API).run
	/home/kishan/code/ipfs-cluster/api/rest/restapi.go:449 +0x8a
FAIL	github.com/ipfs/ipfs-cluster/api/rest	0.094s

```

Though, looking at the logs it seems like it is happening from go-eventbus

After that tests ran fine.

--- a/api/rest/config_test.go
+++ b/api/rest/config_test.go
@@ -177,6 +177,7 @@ func TestLibp2pConfig(t *testing.T) {
 	cfg.ID = pid
 	cfg.PrivateKey = priv
 	addr, _ := ma.NewMultiaddr("/ip4/127.0.0.1/tcp/0")
+	cfg.HTTPListenAddr = addr
 	cfg.Libp2pListenAddr = addr
 
 	err = cfg.Validate()

--- a/api/rest/restapi.go
+++ b/api/rest/restapi.go
@@ -158,13 +158,13 @@ func NewAPIWithHost(ctx context.Context, cfg *Config, h host.Host) (*API, error)
 	api.addRoutes(router)
 
 	// Set up api.httpListener if enabled
-	err = api.setupHTTP(ctx)
+	err = api.setupHTTP()
 	if err != nil {
 		return nil, err
 	}
 
 	// Set up api.libp2pListener if enabled
-	err = api.setupLibp2p(ctx)
+	err = api.setupLibp2p()
 	if err != nil {
 		return nil, err
 	}
@@ -177,7 +177,7 @@ func NewAPIWithHost(ctx context.Context, cfg *Config, h host.Host) (*API, error)
 	return api, nil
 }
 
-func (api *API) setupHTTP(ctx context.Context) error {
+func (api *API) setupHTTP() error {
 	if api.config.HTTPListenAddr == nil {
 		return nil
 	}
@@ -200,12 +200,15 @@ func (api *API) setupHTTP(ctx context.Context) error {
 	return nil
 }
 
-func (api *API) setupLibp2p(ctx context.Context) error {
+func (api *API) setupLibp2p() error {
 	// Make new host. Override any provided existing one
 	// if we have config for a custom one.
 	if api.config.Libp2pListenAddr != nil {
+		// We use a new host context. We will call
+		// Close() on shutdown(). Avoids things like:
+		// https://github.com/ipfs/ipfs-cluster/issues/853
 		h, err := libp2p.New(
-			ctx,
+			context.Background(),
 			libp2p.Identity(api.config.PrivateKey),
 			libp2p.ListenAddrs([]ma.Multiaddr{api.config.Libp2pListenAddr}...),
 		)
@@ -452,7 +455,11 @@ func (api *API) run(ctx context.Context) {
 // runs in goroutine from run()
 func (api *API) runHTTPServer(ctx context.Context) {
 	defer api.wg.Done()
-	<-api.rpcReady
+	select {
+	case <-api.rpcReady:
+	case <-api.ctx.Done():
+		return
+	}
 
 	logger.Infof("REST API (HTTP): %s", api.config.HTTPListenAddr)
 	err := api.server.Serve(api.httpListener)
@@ -464,7 +471,12 @@ func (api *API) runHTTPServer(ctx context.Context) {
 // runs in goroutine from run()
 func (api *API) runLibp2pServer(ctx context.Context) {
 	defer api.wg.Done()
-	<-api.rpcReady
+
+	select {
+	case <-api.rpcReady:
+	case <-api.ctx.Done():
+		return
+	}
 
 	listenMsg := ""
 	for _, a := range api.host.Addrs() {
@@ -496,6 +508,7 @@ func (api *API) Shutdown(ctx context.Context) error {
 
 	api.cancel()
 	close(api.rpcReady)
+
 	// Cancel any outstanding ops
 	api.server.SetKeepAlivesEnabled(false)
 
@@ -506,12 +519,12 @@ func (api *API) Shutdown(ctx context.Context) error {
 		api.libp2pListener.Close()
 	}
 
+	api.wg.Wait()
+
 	// This means we created the host
 	if api.config.Libp2pListenAddr != nil {
 		api.host.Close()
 	}
-
-	api.wg.Wait()
 	api.shutdown = true
 	return nil
 }

